src/main/java/build/buildfarm/instance/memory/MemoryInstance.java

```diff
@@ -321,17 +332,18 @@ private boolean satisfiesRequirements(Platform platform, Operation operation) {
   public boolean watchOperation(
       String operationName,
       boolean watchInitialState,
-      Consumer<Operation> watcher) {
-    List<Consumer<Operation>> operationWatchers = watchers.get(operationName);
+      Function<Operation, Boolean> watcher) {
+    if (watchInitialState &&
+        !watcher.apply(getOperation(operationName))) {
+      return false;
+    }
+    List<Function<Operation, Boolean>> operationWatchers = watchers.get(operationName);
     if (operationWatchers == null) {
       if (watchInitialState) {
-        watcher.accept(null);
+        boolean ignored = watcher.apply(null);
```

Comment:
This was an attempt to suppress a return ignored error, which I've thought better of.  When we can't find the operation, the watch will simply fail with no callback invocation.